src/modules/iotjs_module_mqtt.c

```diff
@@ -0,0 +1,719 @@
+    keep_alive_msb = (uint8_t)(len >> 8);
+    keep_alive_lsb = (uint8_t)(len & 0x00FF);
+  }
+
+
+  iotjs_bufferwrap_t *client_id = iotjs_bufferwrap_from_jbuffer(jclient_id);
+
+  unsigned char variable_header_protocol[7];
+  variable_header_protocol[0] = 0;
+  variable_header_protocol[1] = 4;
+  variable_header_protocol[2] = 'M';
+  variable_header_protocol[3] = 'Q';
+  variable_header_protocol[4] = 'T';
+  variable_header_protocol[5] = 'T';
+  variable_header_protocol[6] = 4;
+
+  size_t variable_header_len = sizeof(variable_header_protocol) +
+                               sizeof(connect_flags) + sizeof(keep_alive_lsb) +
+                               sizeof(keep_alive_msb);
+
+  size_t payload_len = client_id->length + IOTJS_MQTT_LSB_MSB_SIZE;
+  uint32_t remaining_length = payload_len + variable_header_len;
+  size_t full_len = sizeof(header_byte) +
+                    get_remaining_length_size(remaining_length) +
+                    variable_header_len + payload_len;
+
+  jerry_value_t jbuff = iotjs_bufferwrap_create_buffer(full_len);
+  iotjs_bufferwrap_t *buffer_wrap = iotjs_bufferwrap_from_jbuffer(jbuff);
+
+  uint8_t *buff_ptr = (uint8_t *)buffer_wrap->buffer;
+
+  *buff_ptr++ = header_byte;
+
+  buff_ptr = iotjs_encode_remaining_length(remaining_length, buff_ptr);
+
+  memcpy(buff_ptr, variable_header_protocol, sizeof(variable_header_protocol));
+  buff_ptr += sizeof(variable_header_protocol);
+
+  *buff_ptr++ = connect_flags;
+
+  *buff_ptr++ = keep_alive_msb;
+  *buff_ptr++ = keep_alive_lsb;
+
+  buff_ptr = iotjs_mqtt_string_serialize(buff_ptr, client_id);
+
+  jerry_release_value(jclient_id);
+  jerry_release_value(jusername);
+  jerry_release_value(jpassword);
+
+  return jbuff;
+}
+
+
+JS_FUNCTION(MqttPublish) {
+  DJS_CHECK_THIS();
+
+  DJS_CHECK_ARGS(1, object);
+
+  jerry_value_t joptions = JS_GET_ARG(0, object);
+
+  jerry_value_t jmessage =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_MESSAGE);
+  jerry_value_t jtopic =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_TOPIC);
+  jerry_value_t jretain =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_RETAIN);
+  jerry_value_t jqos =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_QOS);
+  jerry_value_t jpacket_id =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_PACKETID);
+
+  uint8_t qos = 0;
+  if (!jerry_value_is_undefined(jqos)) {
+    qos = jerry_get_number_value(jqos);
+  }
+
+  bool dup = false;
+
+  uint8_t header_byte = 0;
+  header_byte |= (PUBLISH << 4);
+  header_byte |= (dup << 3);
+  header_byte |= (qos << 1);
+  header_byte |= (jerry_get_boolean_value(jretain));
+
+  iotjs_bufferwrap_t *message_payload = iotjs_bufferwrap_from_jbuffer(jmessage);
+  iotjs_bufferwrap_t *topic_payload = iotjs_bufferwrap_from_jbuffer(jtopic);
+
+  uint8_t packet_identifier_lsb = 0;
+  uint8_t packet_identifier_msb = 0;
+
+  // IF NOT QOS
+  if (!jerry_value_is_undefined(jqos) &&
```

Comment:
There isa  to_boolean function, which returns with `bool` (never throws error). I would use it here.